scripts/explain-queries.js

```diff
@@ -0,0 +1,121 @@
+//   * Arguments to procedures are always named either `inXxx` or `xxxArg`.
+//   * SQL comment delimiters never appear inside string literals.
+//
+// Non-conformance to those assumptions is not fatal. It just means we'll
+// fail to generate an EXPLAIN for the non-conforming queries.
+
+/* eslint-disable indent, no-console */
+
+'use strict'
+
+const fs = require('fs')
+
+const CREATE_PROCEDURE = /^CREATE PROCEDURE `?([A-Z]+_[0-9]+)/i
+const END_PROCEDURE = /^END;$/i
+const SELECT = /^\s*SELECT/i
+const COMMENT = /--.+$/
+
+const { argv } = process
+
+switch (argv.length) {
+  case 3:
+  case 4:
+    generateExplains(argv[2], argv[3]).forEach(explain => console.log(explain))
+    break
+
+  default:
+    console.error('Usage: node scripts/explain-queries FILE [PROCEDURE]')
+    process.exit(1)
+}
+
+function generateExplains (path, procedureName) {
+  const src = fs.readFileSync(path, { encoding: 'utf8' })
+  const selects = extractSelects(src, procedureName)
+  return selects.map(select => `EXPLAIN ${replaceArgs(select)}`)
+}
+
+function extractSelects (src, procedureName) {
+  let isProcedure = false, isSelect = false
+  const lines = src.split('\n')
+  return lines
+    .reduce((selects, line) => {
+      line = line.replace(COMMENT, '')
+      if (isProcedure) {
+        if (END_PROCEDURE.test(line)) {
+          isProcedure = isSelect = false
+        } else {
+          if (isSelect) {
+            selects[selects.length - 1] += ` ${line.trim()}`
+          } else if (SELECT.test(line)) {
+            selects.push(line.trim())
+            isSelect = true
+          }
+
+          if (line.indexOf(';') !== -1) {
+            isSelect = false
+          }
+        }
+      } else if (procedureName) {
+        const match = CREATE_PROCEDURE.exec(line)
+        if (match && match.length === 2 && match[1] === procedureName) {
+          isProcedure = true
+        }
+      } else {
+        isProcedure = CREATE_PROCEDURE.test(line)
+      }
+
+      return selects
+    }, [])
+    .map(select => purgeUnbalancedParentheses(select))
+}
+
+function purgeUnbalancedParentheses (select) {
+  const openingCount = select.split('(').length
+  const closingCount = select.split(')').length
+
+  if (openingCount < closingCount) {
+    for (let i = 0; i < closingCount - openingCount; ++i) {
+      const index = select.lastIndexOf(')')
+      select = select.substr(0, index) + select.substr(index + 1)
+    }
+  } else if (openingCount > closingCount) {
+    for (let i = 0; i < openingCount - closingCount; ++i) {
+      const index = select.indexOf('(')
+      select = select.substr(0, index) + select.substr(index + 1)
+    }
+  }
+
+  return select
+}
+
+function replaceArgs (select) {
+  return select
+    .replace(/([ \(]?)`?in((?:[A-Z][A-Za-z]+)+)`?/g, replaceArg)
+    .replace(/([ \(]?)`?([a-z]+(?:[A-Z][A-Za-z]+)*Arg)`?/g, replaceArg)
+}
+
+function replaceArg (match, delimiter, arg) {
+  return `${delimiter}0`
```

Comment:
Careful, MySQL sometimes generates different query plans based on what it can infer about provided constants.  In particular, if you give it a fake value for a primary key field, it will detect that there are no matches for that value and abort the query early.  Compare:

```
mysql> explain select * from sessionTokens where tokenId = 0;
+----+-------------+---------------+------+---------------+------+---------+------+------+-------------+
| id | select_type | table         | type | possible_keys | key  | key_len | ref  | rows | Extra       |
+----+-------------+---------------+------+---------------+------+---------+------+------+-------------+
|  1 | SIMPLE      | sessionTokens | ALL  | PRIMARY       | NULL | NULL    | NULL | 3268 | Using where |
+----+-------------+---------------+------+---------------+------+---------+------+------+-------------+
1 row in set (0.00 sec)

mysql> explain select * from sessionTokens where tokenId = 0x0;
+----+-------------+-------+------+---------------+------+---------+------+------+-----------------------------------------------------+
| id | select_type | table | type | possible_keys | key  | key_len | ref  | rows | Extra                                               |
+----+-------------+-------+------+---------------+------+---------+------+------+-----------------------------------------------------+
|  1 | SIMPLE      | NULL  | NULL | NULL          | NULL | NULL    | NULL | NULL | Impossible WHERE noticed after reading const tables |
+----+-------------+-------+------+---------------+------+---------+------+------+-----------------------------------------------------+
1 row in set (0.00 sec)

mysql> select hex(min(tokenId)) from sessionTokens;
+------------------------------------------------------------------+
| hex(min(tokenId))                                                |
+------------------------------------------------------------------+
| 000A06D50BD89F1E993D89149EEA699654CBF5559456DFED1600E583A509BFDB |
+------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> explain select * from sessionTokens where tokenId = 0x000A06D50BD89F1E993D89149EEA699654CBF5559456DFED1600E583A509BFDB;
+----+-------------+---------------+-------+---------------+---------+---------+-------+------+-------+
| id | select_type | table         | type  | possible_keys | key     | key_len | ref   | rows | Extra |
+----+-------------+---------------+-------+---------------+---------+---------+-------+------+-------+
|  1 | SIMPLE      | sessionTokens | const | PRIMARY       | PRIMARY | 32      | const |    1 | NULL  |
+----+-------------+---------------+-------+---------------+---------+---------+-------+------+-------+
1 row in set (0.00 sec)
```

To get good results here, we may actually need to find real values for `uid`, `tokenId` etc and replace them in for arguments based on name, which is a bit bleh.

Reply:
Of course, we could just make sure we `INSERT` specific known values for certain columns during the population phase, then bank on them being present in the switch statement here.